### ISSUE TITLE
unable to enable old TLS version 1.0 & 1.1

### ISSUE THREAD

**Issue by lolorc** [cyrusimap/cyrus-imapd#2934]
Hi there,

I'm in the process of migrating to cyrus 3.0.12.
it is now compiled againt libssl 1.1.1+ and I've had to add 
```
SSL_CTX_set_min_proto_version(ssl_ctx, TLS1_VERSION);
```
to https://github.com/cyrusimap/cyrus-imapd/blob/master/imap/tls.c#L838

to be able to have tls1.0 and tls1.1 enabled with
```
tls_versions tls1_0 tls1_1 tls1_2 tls1_3
```

it can easily checked with nmap:
```
nmap --script ssl-cert,ssl-enum-ciphers -p 143 imap
```

My hack is not enough, as dovecot has done in https://dovecot.org/pipermail/dovecot/2017-August/109175.html it's necessary to set the minimum tls version to the minimum one specified in tls_versions.

**Comment by dilyanpalauzov** [comment]
Do I read correctly, that with `tls_versions tls1_0 tls1_1 tls1_2 tls1_3` you have enabled TLS 1.3 and TLS 1.2, yet not TLS 1.0 and TLS 1.1?  

Per https://github.com/openssl/openssl/blob/95aec8641fb6f596f170f97852e1e2839f4ec2fc/INSTALL openssl can be compiled/configured with no-tls1_1 no-tls1 and then openssl (and all libraries that use it) do not offer TLS 1.0/TLS 1.1.

Do other software on your system that utilizes the same openssl libraries, offer TLS 1.0?

**Comment by lolorc** [comment]
@dilyanpalauzov yes, I'm running Debian Buster, postfix is compiled against the same libssl and it provides tls1.0 & tls 1.1 and as far as I know postfix also uses SSL_CTX_set_min_proto_version.

 I've been able to enable tls 1.0 & 1.1 just by recompiling imapd with 'SSL_CTX_set_min_proto_version(ssl_ctx, TLS1_VERSION);'

**Comment by dilyanpalauzov** [comment]
Reading the documentation of [SSL_CTX_get_min_proto_version()](https://www.openssl.org/docs/man1.1.1/man3/SSL_CTX_get_min_proto_version.html) for me is unclear what happens, when this function is not called.

I filled https://github.com/openssl/openssl/issues/10584.

**Comment by lolorc** [comment]
Thanks,
my mistake, it's just configurable with openssl config.
I'll rename the issue to something else, I guess it's still worth using SSL_CTX_get_min_proto_version() and SSL_CTX_get_max_proto_version() accordingly with the values of tls_version.

**Comment by dilyanpalauzov** [comment]
What means it is configurable with openssl config?

**Comment by lolorc** [comment]
oh pardon my french :)
Debian ships openssl with /etc/ssl/openssl.cnf having
```
[default_conf]
ssl_conf = ssl_sect

[ssl_sect]
system_default = system_default_sect

[system_default_sect]
MinProtocol = TLSv1.2
CipherString = DEFAULT@SECLEVEL=2
```
see my comment in https://github.com/openssl/openssl/issues/10584

**Comment by lolorc** [comment]
hence the usefulness of SSL_CTX_get_min_proto_version, cyrus-imapd should use the tls versions from its config rather than from a default system config.

**Comment by dilyanpalauzov** [comment]
Per https://github.com/openssl/openssl/issues/10584 the preferred way to set the version is with SSL_CTX_set_min_proto_version().  Passing SSL_NO_OP_TLSv1_1 to SSL_CTX_set_options is considered obsolete.

This means that, keeping backward compatibility, the min and max versions shall be extracted from the tls_versions configuration in imapd.conf and passed to SSL_CTX_set_min/max_proto_versions()

**Comment by kroeckx** [comment]
My recommendation is that by default SSL_CTX_set_min_proto_version() should not get called, and the system defaults should get used, but that there should be a way for the users to override it in the config. Postfix is mentioned above, but that is for SMTP, which uses opportunistic encryption. For imap the defaults that Debian sets in the openssl config should be fine for most users, and it should be up to the admin to decide that he still supports users that are limited to TLS 1.0.

**Comment by lolorc** [comment]
@kroeckx, relying on the system defaults mean it's not possible to have a fine default of TLS1.2 for the system and TLS1.0 for imapd while dealing with the huge number of old MUAs.

How could cyrus admins override the system value without calling SSL_CTX_set_min_proto_version() ?

**Comment by kroeckx** [comment]
What I'm saying is that the default config should not have as
effect that SSL_CTX_set_min_proto_version() is called, but that if
the admin does write something in the config to enable TLS 1.0, it
should call SSL_CTX_set_min_proto_version().

**Comment by lolorc** [comment]
@kroeckx oh sorry, I see.
it should not be too difficult to handle min and max, I'll give a proper patch a try as soon as I have more time.

**Comment by ksmurchison** [comment]
Was a pull request with a fix ever opened?
